plugins/inputs/eventhub/eventhub.go

```diff
@@ -0,0 +1,381 @@
+	}
+
+	for i := range metrics {
+		for _, field := range e.ApplicationPropertyFields {
+			if val, ok := event.Get(field); ok {
+				metrics[i].AddField(field, val)
+			}
+		}
+
+		for _, tag := range e.ApplicationPropertyTags {
+			if val, ok := event.Get(tag); ok {
+				metrics[i].AddTag(tag, fmt.Sprintf("%v", val))
+			}
+		}
+
+		if e.SequenceNumberField != "" {
+			metrics[i].AddField(e.SequenceNumberField, *event.SystemProperties.SequenceNumber)
+		}
+
+		if e.EnqueuedTimeAsTs {
+			metrics[i].SetTime(*event.SystemProperties.EnqueuedTime)
+		} else if e.EnqueuedTimeField != "" {
+			metrics[i].AddField(e.EnqueuedTimeField, (*event.SystemProperties.EnqueuedTime).UnixNano()/int64(time.Millisecond))
+		}
+
+		if e.OffsetField != "" {
+			metrics[i].AddField(e.OffsetField, *event.SystemProperties.Offset)
+		}
+
+		if event.SystemProperties.PartitionID != nil && e.PartitionIDTag != "" {
+			metrics[i].AddTag(e.PartitionIDTag, string(*event.SystemProperties.PartitionID))
+		}
+		if event.SystemProperties.PartitionKey != nil && e.PartitionKeyTag != "" {
+			metrics[i].AddTag(e.PartitionKeyTag, *event.SystemProperties.PartitionKey)
+		}
+		if event.SystemProperties.IoTHubDeviceConnectionID != nil && e.IoTHubDeviceConnectionIDTag != "" {
+			metrics[i].AddTag(e.IoTHubDeviceConnectionIDTag, *event.SystemProperties.IoTHubDeviceConnectionID)
+		}
+		if event.SystemProperties.IoTHubAuthGenerationID != nil && e.IoTHubAuthGenerationIDTag != "" {
+			metrics[i].AddTag(e.IoTHubAuthGenerationIDTag, *event.SystemProperties.IoTHubAuthGenerationID)
+		}
+		if event.SystemProperties.IoTHubConnectionAuthMethod != nil && e.IoTHubConnectionAuthMethodTag != "" {
+			metrics[i].AddTag(e.IoTHubConnectionAuthMethodTag, *event.SystemProperties.IoTHubConnectionAuthMethod)
+		}
+		if event.SystemProperties.IoTHubConnectionModuleID != nil && e.IoTHubConnectionModuleIDTag != "" {
+			metrics[i].AddTag(e.IoTHubConnectionModuleIDTag, *event.SystemProperties.IoTHubConnectionModuleID)
+		}
+		if event.SystemProperties.IoTHubEnqueuedTime != nil {
+			if e.IotHubEnqueuedTimeAsTs {
+				metrics[i].SetTime(*event.SystemProperties.IoTHubEnqueuedTime)
+			} else if e.IoTHubEnqueuedTimeField != "" {
+				metrics[i].AddField(e.IoTHubEnqueuedTimeField, (*event.SystemProperties.IoTHubEnqueuedTime).UnixNano()/int64(time.Millisecond))
+			}
+		}
+	}
+
+	id := e.acc.AddTrackingMetricGroup(metrics)
+
+	e.tracker.mux.Lock()
+	e.tracker.messages[id] = metrics
+	e.tracker.mux.Unlock()
+
+	return nil
+}
+
+func (e *EventHub) startTracking(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			if len(e.tracker.messages) == 0 { // everything has been delivered
+				return
+			}
+		case DeliveryInfo := <-e.acc.Delivered():
+			if DeliveryInfo.Delivered() {
+				e.tracker.mux.Lock()
+				delete(e.tracker.messages, DeliveryInfo.ID())
+				e.tracker.mux.Unlock()
+			} else {
+				e.tracker.mux.Lock()
+				id := e.acc.AddTrackingMetricGroup(e.tracker.messages[DeliveryInfo.ID()])
```

Comment:
In order to safely re-add the metrics we would need to save a copy of them.  This way if the metric is modified downstream we can re-add the original metric here.

Reply:
Merged pull request